### ISSUE TITLE
Failed getting delivery receipt containing extended final message states

### ISSUE THREAD

**Issue by chscorpio** [opentelecoms-org/jsmpp#9]
Specifically, Mblox extends the final message states (see Table B-2 in Appendix B of the SMPP 3.4 spec) with the following...
BUFFRED
ACKED
FAILED
https://my.mblox.com/docs/sms.html#smpp-acceptance-status-codes

```
Failed getting delivery receipt
org.jsmpp.util.InvalidDeliveryReceiptException: There is an error found when parsing delivery receipt
        at org.jsmpp.bean.DeliveryReceipt.<init>(DeliveryReceipt.java:80)
        at org.jsmpp.util.DefaultDecomposer.deliveryReceipt(DefaultDecomposer.java:394)
        at org.jsmpp.util.DefaultDecomposer.deliveryReceipt(DefaultDecomposer.java:404)
        at org.jsmpp.bean.DefaultDeliveryReceiptStripper.strip(DefaultDeliveryReceiptStripper.java:36)
        at org.jsmpp.bean.DefaultDeliveryReceiptStripper.strip(DefaultDeliveryReceiptStripper.java:27)
        at org.jsmpp.bean.DeliverSm.getDeliveryReceipt(DeliverSm.java:53)
        at org.jsmpp.bean.DeliverSm.getShortMessageAsDeliveryReceipt(DeliverSm.java:40)
        at com.slooce.smpp.SlooceSMPPSession$1.onAcceptDeliverSm(SlooceSMPPSession.java:80)
        at org.jsmpp.session.SMPPSession.fireAcceptDeliverSm(SMPPSession.java:461)
        at org.jsmpp.session.SMPPSession.access$200(SMPPSession.java:94)
        at org.jsmpp.session.SMPPSession$ResponseHandlerImpl.processDeliverSm(SMPPSession.java:480)
        at org.jsmpp.session.state.SMPPSessionBoundRX.processDeliverSm0(SMPPSessionBoundRX.java:109)
        at org.jsmpp.session.state.SMPPSessionBoundTRX.processDeliverSm(SMPPSessionBoundTRX.java:42)
        at org.jsmpp.session.PDUProcessTask.run(PDUProcessTask.java:90)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
        at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.IllegalArgumentException: No enum constant org.jsmpp.util.DeliveryReceiptState.FAILED
        at java.lang.Enum.valueOf(Enum.java:238)
        at org.jsmpp.util.DeliveryReceiptState.getByName(DeliveryReceiptState.java:60)
        at org.jsmpp.bean.DeliveryReceipt.<init>(DeliveryReceipt.java:75)
        ... 16 more
```

My only workaround today is to pull the source and add them to DeliveryReceiptState.java:

```
$ git diff jsmpp/src/main/java/org/jsmpp/util/DeliveryReceiptState.java
diff --git a/jsmpp/src/main/java/org/jsmpp/util/DeliveryReceiptState.java b/jsmpp/src/main/java/org/jsmpp/util/DeliveryReceiptState.java
index accab45..54e6a35 100644
--- a/jsmpp/src/main/java/org/jsmpp/util/DeliveryReceiptState.java
+++ b/jsmpp/src/main/java/org/jsmpp/util/DeliveryReceiptState.java
@@ -48,7 +48,24 @@ public enum DeliveryReceiptState {
     /**
      * REJECTED
      */
-    REJECTD(7);
+    REJECTD(7),
+
+
+
+    /**
+     * BUFFERED BY MBLOX
+     */
+    BUFFRED(8),
+    /**
+     * ACKNOWLEDGED BY MBLOX
+     */
+    ACKED(9),
+    /**
+     * FAILED BY MBLOX
+     */
+    FAILED(10),
+
+    ;

     private int value;


```

**Comment by pmoerenhout** [comment]
Hi Carlos, I will look into the issue.

Regards Pim

**Comment by chscorpio** [comment]
Thanks. Hopefully, you can find a way to allow extended states.

All I can think of right now is to add DeliveryReceipt.extendedFinalStatus as a String that contains the delivery receipt state as received regardless if it matches the default names in the DeliveryReceiptState enum. The DeliveryReceipt.finalStatus can remain as is for backwards compatibility and set to UNKNOWN in the case of a new state. The toString serializer would use the extendedFinalStatus instead of the finalStatus enum name to reconstruct the original message.

Then apps using MBlox could process DeliveryReceipts depending on the DeliveryReceipt.extendedFinalStatus.

**Comment by chscorpio** [comment]
Thanks @pmoerenhout  You resolved this issue for me in v2.3.6 when you provided an example that shows how jsmpp can support a CustomDeliveryReceiptState implementation.

https://github.com/opentelecoms-org/jsmpp/commit/74052af2f238502542df1b0ebdc6a3066f06871e
